File: leetcode/editor/cn/496_下一个更大元素 I_test.py
import unittest
from typing import List


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        ans = self.solution.nextGreaterElement([1, 4, 100],
                                               [1, 10, 4, 5, 2, 100])
        self.assertEqual([10, 5, -1], ans)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def nextGreaterElement(self,
                           nums1: List[int],
                           nums2: List[int]) -> List[int]:
        mapping: dict[int, int] = {}
        stack: list[int] = []
        # reversed()
        for x in reversed(nums2):
            while stack and stack[- 1] <= x:
                stack.pop()
            mapping[x] = stack[- 1] if stack else -1
            stack.append(x)
        return [mapping[x] for x in nums1]


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()

File: leetcode/editor/cn/496_下一个更大元素 I_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when496_thenSuccess(t *testing.T) {
	assert.Equal(t, 1, 1, "shouldEqual")
}

//leetcode submit region begin(Prohibit modification and deletion)
func nextGreaterElement(nums1 []int, nums2 []int) []int {
	mapping := map[int]int{}
	var stack []int
	for i := len(nums2) - 1; i >= 0; i-- {
		n := nums2[i]
		bigger := -1
		for lng := len(stack); lng > 0; lng = len(stack) {
			top := stack[lng-1]
			if top > n {
				bigger = top
				break
			}
			stack = stack[:lng-1]
		}
		stack = append(stack, n)
		mapping[n] = bigger
	}
	var ans []int
	for _, n := range nums1 {
		ans = append(ans, mapping[n])
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)
